### PR TITLE
Added docstring to @fastmath macro.

### PR DESCRIPTION
**Pull request by NHDaly**
Adds docstring to @fastmath macro that explains a bit about what it does, and includes examples how to use it.

Pulled the detailed list of effects from the top of the file.

#doc

### CODE REVIEW

File: base/fastmath.jl [hunk 1]
```diff
@@ -124,6 +124,32 @@ function make_fastmath(symb::Symbol)
...
+
+This makes the following transformations:
+- nnan: No NaNs - Allow optimizations to assume the arguments and
+        result are not NaN.
+- ninf: No Infs - Allow optimizations to assume the arguments and
+        result are not +/-Inf.
+- nsz:  No Signed Zeros - Allow optimizations to treat the sign of a
+        zero argument or result as insignificant.
+- arcp: Allow Reciprocal - Allow optimizations to use the reciprocal
+        of an argument rather than perform division.
```
Comment (yuyichao) [inline base/fastmath.jl]: Assuming you are copying from langref, the transformation allowed is much broader than this list.

**Review by yuyichao — CHANGES_REQUESTED** [review]

File: base/fastmath.jl (hunk 1 above)
Comment (yuyichao) [inline base/fastmath.jl]: Ah, I see that you pull the list from the comment, in that case, maybe you should use http://llvm.org/docs/LangRef.html#fast-math-flags

Comment (NHDaly) [inline base/fastmath.jl]: Oh, thanks! I've updated with a second commit. I can squash before merging.

Does this look better? I added a link to the llvm docs (and added one at the top of the file), and I listed the missing flags.

Comment (ararslan) [inline base/fastmath.jl]: No need to manually squash; we can squash the commits on merge.

**Review by yuyichao — APPROVED** [review]
I'd prefer shrinking this a little (i.e. `@fastmath` allows all sorts of transformation and then mention the few specific ones as special cases) but this looks correct and complete enough to me....

File: base/fastmath.jl [hunk 2]
```diff
@@ -124,6 +125,38 @@ function make_fastmath(symb::Symbol)
 end
 make_fastmath(expr) = expr
 
+"""
+@fastmath expr
```
Comment (yuyichao) [inline base/fastmath.jl]: Actually, this should be a code block (indent it) and this should be added to the doc/stdlib/

**Review by yuyichao — CHANGES_REQUESTED** [review]

File: base/fastmath.jl (hunk 2 above)
Comment (NHDaly) [inline base/fastmath.jl]: >Actually, this should be a code block (indent it) 

Just indent everything inside the triple quotes, you mean?

> and this should be added to the doc/stdlib/

Sorry, I don't understand what you mean. Like, there should be a page for this in the official docs? Like a new markdown file?

Comment (yuyichao) [inline base/fastmath.jl]: > Just indent everything inside the triple quotes, you mean?

No, just the signature, see [the doc for other macros/functions](https://github.com/JuliaLang/julia/blob/d3db31200c09a99ad9f7cbc019dd0aa56c6944a7/base/essentials.jl#L200)

> added to the doc/stdlib/

See https://github.com/JuliaLang/julia/blob/d3db31200c09a99ad9f7cbc019dd0aa56c6944a7/CONTRIBUTING.md#improving-documentation There are examples there that you can follow. And you can decide which file to put it in...

Comment (NHDaly) [inline base/fastmath.jl]: >No, just the signature, see the doc for other macros/functions

Gotcha.  Done. thanks!

See [CONTRIBUTING.md#improving-documentation](https://github.com/JuliaLang/julia/blob/d3db31200c09a99ad9f7cbc019dd0aa56c6944a7/CONTRIBUTING.md#improving-documentation)

Okay thanks. I made those changes, and I'm running `make docs` now. I'll post again when that's done.

Comment (NHDaly) [inline base/fastmath.jl]: oh noes! Jk i can't build on this computer; i don't have all the dependencies installed. I'll push this tomorrow. Thanks! :)

Comment (yuyichao) [inline base/fastmath.jl]: Note that `make docs` is only needed for testing. There shouldn't be anything that you need to push after that.

**Review by yuyichao — APPROVED** [review]

**Comment by ararslan** [comment]
Thanks for the contribution! For future reference, it's a good idea to reference macros in commit messages as e.g. `at-fastmath rather than `@fastmath`. In the latter case, GitHub will send a notification to the user called fastmath. Whoever squash-merges this can edit the commit message here appropriately.

**Comment by NHDaly** [comment]
> I'd prefer shrinking this a little (i.e. @fastmath allows all sorts of transformation and then mention the few specific ones as special cases) but this looks correct and complete enough to me....

@yuyichao I agree. What about just dropping everything after the definition, and dropping `fast` since it's so vague, so it looks like this:

```
"""
    @fastmath expr

Executes a transformed version of the expression, which calls
functions that may violate strict IEEE semantics. This allows
the fastest possible operation, but results are undefined.

This sets the [LLVM Fast-Math
flags](http://llvm.org/docs/LangRef.html#fast-math-flags), including
the following transformations:
- nnan: No NaNs
- ninf: No Infs
- nsz:  No Signed Zeros
- arcp: Allow Reciprocal
- contract: Allow floating-point contraction.

# Examples
```jldoctest
julia> @fastmath 1+2
3

julia> @fastmath(sin(3))
0.1411200080598672
```
"""
```

**Comment by NHDaly** [comment]
> For future reference, it's a good idea to reference macros in commit messages as e.g. `at-fastmath` rather than `@fastmath`. In the latter case, GitHub will send a notification to the user called fastmath. Whoever squash-merges this can edit the commit message here appropriately.

:D haha that makes a lot of sense. Thanks!! I'll let you do the change so my rebase doesn't make the PR harder to read.

Cheers!
~Nathan

**Comment by yuyichao** [comment]
> dropping fast since it's so vague,

I'll not drop the `fast` description though it'll be better to drop those flag names altogether, maybe only with a mention that this corresponds to the llvm `fast` flag. There are many more transformations allowed for `fast` that's not included in those categories.

File: base/fastmath.jl (hunk 2 above)
Comment (NHDaly) [inline base/fastmath.jl]: Gotcha. I have verified the docs build and look correct.

**Comment by NHDaly** [comment]
> I'll not drop the fast description though it'll be better to drop those flag names altogether, maybe only with a mention that this corresponds to the llvm fast flag. There are many more transformations allowed for fast that's not included in those categories.

Okay, I've pushed a new version that removes the detailed bullets and instead links to the following page in the docs:
https://docs.julialang.org/en/stable/manual/performance-tips#performance-annotations

Is there a better way to link to docs, such that it will keep you in the correct, corresponding version? (ie don't always link to `stable`?)

**Comment by fredrikekre** [comment]
> Is there a better way to link to docs

You can use `@ref` if you add an `@id` to the section header. Something like this:
```patch
diff --git a/base/fastmath.jl b/base/fastmath.jl
index f6be79a..6f925e4 100644
--- a/base/fastmath.jl
+++ b/base/fastmath.jl
@@ -133,7 +133,10 @@ functions that may violate strict IEEE semantics. This allows
 the fastest possible operation, but results are undefined.
 
 This sets the [LLVM Fast-Math
-flags](http://llvm.org/docs/LangRef.html#fast-math-flags), and corresponds to the `-ffast-math` option in clang. See [the notes on performance annotations](https://docs.julialang.org/en/stable/manual/performance-tips#performance-annotations) for more details.
+flags](http://llvm.org/docs/LangRef.html#fast-math-flags), and corresponds to the
+`-ffast-math` option in clang.
+See [the notes on performance annotations](@ref man-performance-annotations) for
+more details.
 
 # Examples
 ```jldoctest
diff --git a/doc/src/manual/performance-tips.md b/doc/src/manual/performance-tips.md
index d171cbf..02f609b 100644
--- a/doc/src/manual/performance-tips.md
+++ b/doc/src/manual/performance-tips.md
@@ -1082,7 +1082,7 @@ These are some minor points that might help in tight inner loops.
   * Use [`div(x,y)`](@ref) for truncating division of integers instead of [`trunc(x/y)`](@ref), [`fld(x,y)`](@ref)
     instead of [`floor(x/y)`](@ref), and [`cld(x,y)`](@ref) instead of [`ceil(x/y)`](@ref).
 
-## Performance Annotations
+## [Performance Annotations](@id man-performance-annotations)
 
 Sometimes you can enable better optimization by promising certain program properties.

```

File: base/fastmath.jl [hunk 3]
```diff
@@ -124,6 +125,25 @@ function make_fastmath(symb::Symbol)
...
 
+"""
+    @fastmath expr
+
+Executes a transformed version of the expression, which calls
+functions that may violate strict IEEE semantics. This allows
+the fastest possible operation, but results are undefined.
+
+This sets the [LLVM Fast-Math
+flags](http://llvm.org/docs/LangRef.html#fast-math-flags), and corresponds to the `-ffast-math` option in clang. See [the notes on performance annotations](https://docs.julialang.org/en/stable/manual/performance-tips#performance-annotations) for more details.
```
Comment (StefanKarpinski) [inline base/fastmath.jl]: wrap 92 please

Comment (NHDaly) [inline base/fastmath.jl]: Oops. Done, thanks.

**Comment by NHDaly** [comment]
PTAL: I think this is all set, and the docs build correctly. Thanks again for your reviewing!

> You can use @ref if you add an @id to the section header. Something like this:

Thanks, done.

**Comment by yuyichao** [comment]
Be careful to code quote all `@`s that are not intended to mention people, even in text quote.....

I did not know that `ref` and `id` are real github users.... (with `id` actually seems to be active)

**Comment by NHDaly** [comment]
Haha wow that's hard. I'll be sure to be careful. I didn't do it again, did I?

**Comment by yuyichao** [comment]
> I didn't do it again, did I?

I edited the comment to fix it.

**Review by ararslan — APPROVED** [review]
LGTM. Thanks for the contribution!

File: base/fastmath.jl [hunk 4]
```diff
@@ -124,6 +125,27 @@ function make_fastmath(symb::Symbol)
 end
 make_fastmath(expr) = expr
 
+"""
+    @fastmath expr
+
+Executes a transformed version of the expression, which calls functions that
```
Comment (Sacha0) [inline base/fastmath.jl]: Perhaps `Executes` -> `Execute` in keeping with the documentation guidelines? :)

**Review by Sacha0 — APPROVED** [review]
lgtm! Thanks @NHDaly! :)

File: base/fastmath.jl (hunk 4 above)
Comment (NHDaly) [inline base/fastmath.jl]: Done, thanks! Sorry about that.

**Review by ararslan — APPROVED** [review]

**Comment by Sacha0** [comment]
Thanks for the great contribution @NHDaly ! :)

**Comment by fastmath** [comment]
Thank you for documenting me.

**Pull request merged**